api/slideshow/api.go

```diff
@@ -0,0 +1,148 @@
+package slideshow
+
+import (
+	"errors"
+	"log"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/pisign/pisign-backend/types"
+	"github.com/pisign/pisign-backend/utils"
+)
+
+type API struct {
+	types.BaseAPI
+	FilePathsForTags []string
+
+	Config types.SlideShowConfig
+	// Add other fields as necessary, (keep lowercase to avoid being stored in json)
+}
+
+// NewAPI creates a new API
+func NewAPI(sockets map[types.Socket]bool, pool types.Pool, id uuid.UUID) *API {
+	a := new(API)
+	a.BaseAPI.Init("slideshow", sockets, pool, id)
+
+	return a
+}
+
+// Configure based on client message
+func (a *API) Configure(message types.ClientMessage) error {
+	// Make sure the client widget is immediately sent new data with new config options
+	defer func() {
+		if a.Pool != nil && a.Sockets != nil {
+			a.Pool.Save()
+			a.Send(a.Data())
+		}
+	}()
+
+	if err := a.BaseAPI.Configure(message); err != nil {
+		return err
+	}
+
+	switch message.Action {
+	case types.ConfigureAPI, types.Initialize:
+		log.Printf("Configuring Slideshow:\n %s\n", string(message.Config))
+		if err := utils.ParseJSON(message.Config, &a.Config); err != nil {
+			log.Println("Could not properly configure Slideshow")
+			return errors.New("could not properly configure Slideshow")
+		}
+
+		// Add custom checks for config fields here (see the `time` api as an example)
+
+		log.Println("Slideshow configuration successful:", a)
+	}
+	log.Println("Slideshow config set to ", a.Config)
+
+	return nil
+}
+
+// Data performs a bulk of the computational logic
+func (a *API) Data() (interface{}, error) {
+	// Get the filepaths to include for a set of tags
+	// Also send all the possible tags so the client can choose later which tags to use
+	// when configuring the api
+	imgDB := a.Pool.GetImageDB().(*types.ImageDB)
+	images := make([]string, imgDB.NumImages-1)
```

Comment:
Why is the `-1` necessary?

Reply:
It was making it too big, and adding extra empty things in it D:

Reply:
Hmmm I feel like that's a symptom of a worse problem in the ImageDB somewhere. Maybe the `NumImages` field is out of sync with the actual number of images? Maybe we could turn `NumImages` into a function like this:
```go
func (imgDB *ImageDB) NumImages() int {
    return len(imgDB.Images)
}
```

This way it's always in sync